src/main/java/com/google/edith/services/SearchServiceImpl.java

```diff
@@ -0,0 +1,152 @@
+import com.google.common.collect.ImmutableList;
+import com.google.edith.interfaces.SearchService;
+import com.google.edith.servlets.Item;
+import com.google.edith.servlets.Receipt;
+import java.util.ArrayList;
+import java.util.List;
+
+/** Service to create query and handle searching on entities. */
+public final class SearchServiceImpl implements SearchService {
+
+  private final DatastoreService datastore;
+  private final UserService userService;
+
+  public SearchServiceImpl(DatastoreService datastore, UserService userService) {
+    this.datastore = datastore;
+    this.userService = userService;
+  }
+
+  @Override
+  public ImmutableList<Receipt> createReceiptObjects(ImmutableList<Entity> entities) {
+    List<Receipt> receipts = new ArrayList<>();
+
+    for (Entity entity : entities) {
+      Key entityKey = entity.getKey();
+      Query itemQuery = new Query("Item", entityKey);
+      PreparedQuery results = datastore.prepare(itemQuery);
+      List<Entity> itemEntities = results.asList(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
+      ImmutableList<Item> items = createItemObjects(ImmutableList.copyOf(itemEntities));
+      Receipt receipt =
+          new Receipt(
+              (String) entity.getProperty("userId"),
+              (String) entity.getProperty("storeName"),
+              (String) entity.getProperty("date"),
+              (String) entity.getProperty("name"),
+              (String) entity.getProperty("fileUrl"),
+              // Datastore returns as double even when uploaded as float.
+              (float) ((double) entity.getProperty("price")),
+              items.toArray(new Item[items.size()]));
+
+      receipts.add(receipt);
+    }
+    return ImmutableList.copyOf(receipts);
+  }
+
+  @Override
+  public ImmutableList<Item> createItemObjects(ImmutableList<Entity> entities) {
+    List<Item> itemsList = new ArrayList<>();
+
+    for (Entity entity : entities) {
+      Item item =
+          Item.builder()
+              .setUserId((String) entity.getProperty("userId"))
+              .setName((String) entity.getProperty("name"))
+              .setPrice((double) entity.getProperty("price"))
+              .setQuantity((long) entity.getProperty("quantity"))
+              .setDate((String) entity.getProperty("date"))
+              .setCategory((String) entity.getProperty("category"))
+              .setExpiration((String) entity.getProperty("expireDate"))
+              .build();
+
+      itemsList.add(item);
+    }
+    return ImmutableList.copyOf(itemsList);
+  }
+
+  @Override
+  public ImmutableList<Entity> findEntityFromDatastore(
+      String name, String date, String kind, String sortOrder, String sortOnProperty) {
+    Query query = prepareQuery(name, date, kind, sortOrder, sortOnProperty);
+    PreparedQuery results = datastore.prepare(query);
+    List<Entity> entities = results.asList(FetchOptions.Builder.withLimit(Integer.MAX_VALUE));
+    return ImmutableList.copyOf(entities);
+  }
```

Comment:
This method is a bit unwieldy with so many params and it would be ideal to break up the method into each param so the caller can directly use findEntityWithName() rather than using this method and giving a bunch of empty strings. You can add a todo if thats easier :)

Reply:
Sure. Thanks for the suggestion.